lib/policies/proxy/proxy.js

```diff
@@ -51,15 +54,49 @@ module.exports = function (params, config) {
   if (endpoint.url) {
     return function proxyHandler (req, res) {
       logger.debug(`proxying to ${endpoint.url}, ${req.method} ${req.url}`);
-      proxy.web(req, res, { target: endpoint.url });
+      proxy.web(req, res, proxyOptions);
     };
   }
 
   // multiple urls will load balance, defaulting to round-robin
   const strategy = params.strategy || 'round-robin';
-  const balancer = createStrategy(strategy, proxy, endpoint.urls);
+  const balancer = createStrategy(strategy, proxy, proxyOptions, endpoint.urls);
 
   return function balancerProxyHandler (req, res) {
     return balancer.proxy(req, res);
   };
 };
+
+// Parse endpoint URL if single URL is provided.
+// Extend proxy options by allowing and parsing keyFile and certFile.
+function parseProxyOptions (proxyOptions = {}, endpoint = {}) {
+  let parsedProxyOptions = {};
+
+  if (Object.keys(endpoint).includes('proxyOptions')) {
+    parsedProxyOptions = Object.assign(endpoint.proxyOptions, proxyOptions);
+  }
+
+  if (!Object.keys(parsedProxyOptions).includes('target')) parsedProxyOptions.target = {};
+
+  if (parsedProxyOptions.target.keyFile) {
+    parsedProxyOptions.target.key = fs.readFileSync(parsedProxyOptions.target.keyFile);
+    delete parsedProxyOptions.target.keyFile;
+  }
+
+  if (parsedProxyOptions.target.certFile) {
+    parsedProxyOptions.target.cert = fs.readFileSync(parsedProxyOptions.target.certFile);
+    delete parsedProxyOptions.target.certFile;
+  }
+
+  if (parsedProxyOptions.target.caFile) {
+    parsedProxyOptions.target.ca = fs.readFileSync(parsedProxyOptions.target.caFile);
+    delete parsedProxyOptions.target.caFile;
+  }
+
+  if (endpoint.url) {
+    const parsedEndpointUrl = endpoint.url ? url.parse(endpoint.url) : {};
+    parsedProxyOptions.target = Object.assign(parsedProxyOptions.target, parsedEndpointUrl);
+  }
+
+  return parsedProxyOptions;
```

Comment:
Using ``` delete ``` is not required, but is keeping the object being manipulated clean. Hence doing it.

What would you prefer?

Reply:
@srcnix @XVincentX 
xxxxFile is just some string. it is not a lot of data. why we cannot just remove the delete part and allow passing everything? 

@XVincentX  creating new object will require either listing all properties or using some `_.omit` it is kind of performance impact and code readabliity  issue

Reply:
The thing here is that, in the current state, `parseProxyOptions` function is not pure, because it's modifying the input argument `endpoints`.

This side effect makes the function impossible to be called twice with the same arguments (because properties are being removed) from the input object.

The problematic line is `parsedProxyOptions = Object.assign(endpoint.proxyOptions, proxyOptions);`
(as the returned reference is on `endpoint.proxyOptions`)

Either we create a new, parsed object and return it, or let's do something to make sure we do not modify user provided arguments.

Reply:
Gotcha! Shall update.